### ISSUE TITLE
error CL INVALID KERNEL when calling clEnqueueNDRangeKernel for kernel 0

### ISSUE THREAD

**Issue by tremendo88** [fireice-uk/xmr-stak#1799]
start the miner, everything starts to perfection after an undefined time (about 5 minutes every hour) the miner seems to stop, I return this error: **error CL INVALID KERNEL when calling clEnqueueNDRangeKernel for kernel 0**
After about 5 minutes it resumes to regularly mine. While the GPU temperature is blocked, it drops considerably, while the minion signals hash millionaires. It starts again after only 5 minutes. But the fact that you stop 5 minutes every hour is a big loss. I have increased the virtual memory to 80000 but it keeps giving me the same problem. How can I solve this problem?

# Basic information
  - Intel GPU G4400.
  - GPU rx 580 nitro+ 8GB memory Micron.

# Compile issues
  - OS Win 10 LTSB
  ```
 **all** 
[2018-08-21 23:55:27] : Error CL_INVALID_KERNEL when calling clEnqueueNDRangeKernel for kernel 0.
[2018-08-21 23:55:27] : Error CL_INVALID_KERNEL when calling clEnqueueNDRangeKernel for kernel 0.
[2018-08-21 23:55:27] : Error CL_INVALID_KERNEL when calling clEnqueueNDRangeKernel for kernel 0.
[2018-08-21 23:55:27] : Error CL_INVALID_KERNEL when calling clEnqueueNDRangeKernel for kernel 0.
[2018-08-21 23:55:27] : Error CL_INVALID_KERNEL when calling clEnqueueNDRangeKernel for kernel 0.
[2018-08-21 23:55:27] : Error CL_INVALID_KERNEL when calling clEnqueueNDRangeKernel for kernel 0.
[2018-08-21 23:55:27] : Error CL_INVALID_KERNEL when calling clEnqueueNDRangeKernel for kernel 0.
[2018-08-21 23:55:27] : Switching back to user pool.
HASHRATE REPORT - AMD
| ID |    10s |    60s |    15m | ID |    10s |    60s |    15m |
|  0 | 2030238.3 | 2035735.9 | 270105.1 |  1 | 2058242.8 | 2053430.3 | 269392.8 |
|  2 | 3164187.0 | 3169292.3 | 420170.2 |  3 | 3170461.5 | 3183978.1 | 428960.5 |
Totals (AMD):  10423129.6 10442436.7 1388628.6 H/s
-----------------------------------------------------------------
Totals (ALL):   10423129.6 10442436.7 1388628.6 H/s
Highest:  11658101.3 H/s
-----------------------------------------------------------------
[2018-08-21 23:55:37] : Result accepted by the pool.
HASHRATE REPORT - AMD
| ID |    10s |    60s |    15m | ID |    10s |    60s |    15m |
|  0 |  454.0 | 1686942.3 | 270254.3 |  1 |  446.4 | 1695867.8 | 269444.9 |
|  2 |  454.0 | 2646218.2 | 420697.9 |  3 |  451.9 | 2646550.4 | 428580.3 |
Totals (AMD):  1806.3 8675578.7 1388977.4 H/s
-----------------------------------------------------------------
Totals (ALL):   1806.3 8675578.7 1388977.4 H/s
Highest:  11658101.3 H/s
-----------------------------------------------------------------
 **full** 
  ```

  ```
  run `cmake -LA .` in the build folder and add the output here
  ```

# Issue with the execution
  - Do you compiled the miner by our own?
  ```
  run `./xmr-stak --version-long` and add the output here
  ```

# AMD OpenCl issue

  ```
  run `clinfo` and add the output here
  ```

# Stability issue

this is my overclock

GPU  1250 / 900mV and 2200 / 900mV 
// gpu: Ellesmere memory:3920
  // compute units: 36
  { "index" : 0,
    "intensity" : 896, "worksize" : 8,
    "affine_to_cpu" : false, "strided_index" : 1, "mem_chunk" : 2,
    "comp_mode" : true
  },
  // gpu: Ellesmere memory:3920
  // compute units: 36
  { "index" : 0,
    "intensity" : 896, "worksize" : 8,
    "affine_to_cpu" : false, "strided_index" : 1, "mem_chunk" : 2,
    "comp_mode" : true
  },

**Comment by rdugan** [comment]
Just experienced this as well...  The reason it's 5 minutes every hour is because it's happening during the donation window - the behavior stopped as soon as the miner reported 'Switching back to user pool'.

I'm using the cryptonight_bittube2 algo - i'm guessing the donation window is attempting to use an algo mismatched to the current settings?

**Comment by aguerragza** [comment]
This is annoying has anybody fixed this

**Comment by Spudz76** [comment]
be sure to clear out the `.openclcache` or use the command line option to disable cached kernels and make sure it has the fresh code matched to your current driver

try with strided_index:0 instead (flat memory)

The algo switching probably can't do strided (and maybe not even flat, but it's the most likely to work OK) because bittube2 is 4MB blocksize (CN-Heavy) while donation is as far as I know always 2MB (normal CN).  It should have built two kernels at the start one for the main pool algo and one for the donation pool algo, and it is having trouble calling the second one, or it is operating on the wrong memory locations due the skew.

Compile with donation disabled for an easy permanent fix, send some manually to the addresses in the main readme instead.

**Comment by rdugan** [comment]
The behavior of switching algos for donations is just awful - period.  Many miners tune their cards to the limits for the algo their mining, so switching to a different algo can (and usually does) cause major instability.

**Comment by rdugan** [comment]
Wow - thanks for the lecture on hashrate/watt efficiency, but maybe come down off the high horse and think a little outside the box before dumping all over someone trying to help improve the product.  I'm talking about cutting power use to the absolute minimum for any chosen hashrate, and I some people in this community actually think the devs deserve donations, even if we can compile them out.

I would think a 'contributor' to mining software would get both of these concepts.

**Comment by Spudz76** [comment]
OK but when you drive close to edge of the road, you're gonna lose some tires to the curb.

That doesn't change depending on anyones horse heights.

**Comment by rdugan** [comment]
You're gonna lose a lot more tire when someones changing the curb position on you

**Comment by rdugan** [comment]
Are you kidding/trolling?  I started talking about algo changes for donation mining.  I'm still talking about algo changes for donation mining.  Not that hard to follow.

Not to worry - plenty of other miners out there, the devs of which understand this simple concept, and don't berate users for providing feedback.  I'm sure they'll appreciate my donations.  Good luck.

**Comment by Spudz76** [comment]
Donation is really just for the lazy, if you bother to compile just shut off the buggy dual coin donating.
If you are lazy then I guess crash.

You are arguing about a feature you can easily disable, and started it by berating the coin switching action of the donation phase as if it were a mentally deficient idea in the first place.

It should work, your hardware sucks, so compile for your hardware and tune it to not crash on your hardware (remove donation, its five keystrokes or less)

**Comment by rdugan** [comment]
For anyone interested in actually having this issue actually addressed, in a fit of 'laziness', I've migrated all of my hardware that 'sucks' (30+ vegas and 580s) to both JCE and SRB miners.  While both are closed-source, not only have they seen fit to avoid this particular deficiency in their code, they also produce better hashrates.  And as an added bonus, most issues are addressed by reputable, responsive developers, so you can avoid arguing with socially handicapped contributors.

**Comment by psychocrypt** [comment]
Only one note: SRB miner is a fork of xmr-stak and is not providing the
source code. This is a license violation.

**Comment by rdugan** [comment]
I had actually suspected that...  I appreciate your contributions to the community @psychocrypt.  Hope you guys can address this issue so you can keep your donations, as opposed to what has been suggested here.

**Comment by psychocrypt** [comment]
I am thinking very often about how to solve it. The problem is that we can
not mine all coins we support with the dev pool.
Which coin do you mine?

**Comment by Spudz76** [comment]
@rdugan And on my part, apologies for gruff.  Suppose hanging out in support too much makes for troll mood... everyone comes in talking negative while we just try to make a good workable tool (everyone is like "you sabotaged me!" in their tone, but maybe I am just reading it that way somewhat also).

I will try to be more passive even against repeated user-trolling-us haha.

**Comment by rdugan** [comment]
Lately have vegas on the cnv2 algo (xmr), with 580s on heavy/saber/haven.  Occasionally shift things around based on market conditions.

Can you guys possibly at least maintain one wallet for each algo?  Coin switching w/in the same algo shouldn't be an issue.

@Spudz76 - no worries.  and i got a little carried away as well.  Deleting my cross-advert comment - i generally despise that nonsense.

**Comment by psychocrypt** [comment]
The problem is that we would need to maintain 10 different blockchains.
Also exchanging it to a usefull currency would be hard because we need to
tax the transfers. Holding most currencies will also be not an option
because I personally think that most coins will be not live very long.

**Comment by Spudz76** [comment]
donation detour would probably not even be there if people would not 'forget' to pay it forward (backward?) now and then

alas, why governments have taxes not donation boxes

**Comment by rdugan** [comment]
Fair enough.  What about @Spudz76's suggestion of alternate binaries for donations (assuming different algo,) possibly w/ significant (10-20%?) reductions to intensity, etc vs users actual settings.

**Comment by Spudz76** [comment]
I don't care for the "if I can't have donation, you get slow" method like claymore etc uses - its similar to if I can't have it nobody will mentality (abusive/controlling/wasteful for spite?).  Also being open source it would be just as removable as the donation is, though more difficult.  Any binary blob trick would just get ripped like the 80 versions of free-version claymore products (and he put way more antimod stuff in that we would bother)

**Comment by rdugan** [comment]
Oh - i don't mean reducing performance of users chosen algo.  I mean reducing intensity for donation window algo, to guard against exceeding performance envelope of gpu as configured for main algo.

**Comment by Spudz76** [comment]
Ahhhhh, I see.  I thought capped intensity like claymore (or however he hobbles you when you choose not to donate) basically go ahead and put in 1024 but it will be 768 effective...

**Comment by Spudz76** [comment]
And yes, the major issue is the auto intensity does not work for other algos, yes.  I had considered briefly last night splitting the configs into algo-types (multiple sections, more complex for the user, but at least tunable for all possible cointypes at runtime)

**Comment by Spudz76** [comment]
^^ would also allude to the user that different coins need different tunings (not currently super obvious unless you have been a coin wrangler for a while)

**Comment by Spudz76** [comment]
still have a problem of not knowing that a certain intensity on a certain device on a certain OS might not like what math says it should like, so the forked miner config is the only real method to accomplish both some auto-guidance and have user override-adjustability for each algo in case it switches at runtime.  hmm

**Comment by rdugan** [comment]
Yeah - that's why i'm thinking of some sort of equation/model to auto convert between main algo (whether auto or manual) and donation algo, then a haircut on top (assuming auto-config'd donation algo) to be safe. 

An auto-guided multi-algo config could solve the problem AND be useful for an algo-switching pool feature, but should be brain-dead enough to functionally address donation window algo swap issues without requiring manual intervention/setup by single-algo users.
